hadoop-hdfs-project/hadoop-hdfs/src/site/markdown/NamenodeFGL.md

```diff
@@ -0,0 +1,201 @@
+
+For type 2, NN acquires full path locks in a predefined order, such as the lexicographic order, to avoid deadlocks.
+
+For type 3, NN acquires a full path lock by in the following fashion:
+- Unsafely obtains full path recursively
+- Acquires the full path lock according to the lock mode
+- Rechecks whether the last node of the full path is equal to the INodeID given
+  - If not, that means that the `INodeFile` might have been renamed or concatenated, need to retry
+  - If the max retry attempts have been reached, throw a `RetryException` to client to let client retry
+
+### `INodeFile` Lock
+
+Used to protect operations involving blocks.
+
+One block belongs to one and only one `INodeFile`, so NN can use the INodeFile lock to make operations thread-safe. Normally, there is no need to acquire the full path lock since changing the namespace tree structure does not affect the block.
+
+`concat` might change the `INodeFile` a block belongs to. Since both block related operations and `concat` need to acquire the `INodeFile` write lock, only one of them can be processed at a time.
+
+### DN Lock
+
+Used to protect operations involving DNs.
+
+NN uses a `DatanodeDescriptor` object to store the information of one DN and uses `DatanodeManager` to manage all DNs in the memory. `DatanodeDescriptor` uses `DatanodeStorageInfo` to store the information of one storage device on one DN.
+
+DNs have nothing to do with the namespace tree, so NN uses a separate DN lock for these operations. Since DNs are independent of one another, NN can assign a lock to each DN.
+
+### Global Lock
+
+Used for operations requiring the entire namespace locked.
+
+There are some operations that need to lock the entire namespace, e.g. safe mode related operations, HA service related, etc. NN uses the global lock to make these operations thread-safe. Outside of these infrequent operations that require the global write lock, all other operations have to acquire the global read lock.
+
+### Lock Order
+
+As mentioned above, there are the global lock, DN lock, and full path lock. NN acquires locks in this specific order to avoid deadlocks.
+
+Locks are to be acquired in this order:
+- Global > DN > Full path
+- Global > DN > Last `INodeFile`
+
+Possible lock combinations are as follows:
+- Global write lock
+- Global read lock > Full path lock
+- Global read lock > DN read/write lock
+- Global read lock > DN read/write lock > Read/Write lock of last `INodeFile`
+- Global read lock > DN read/write lock > Full path lock
+
+### Lock Pools
+
+NN allocates locks as needed to the INodes used by active threads, and deletes them after the locks are no longer in use. Locks for commonly accessed `INode`s like the root are cached.
+
+NN uses an `INodeLockPool` to manage these locks. The lock pool:
```

Comment:
There is improvement to fix the hot iNodes, right? We can just mention it.

Reply:
Do you mean caching the locks for frequently used INodes? It's already mentioned above